File: HashTable/find_first_unique_char.go
package HashTable

func FindUniqueChar(str string) int {
	m := make(map[string]int)

	for i := 0; i < len(str); i++ {
		char := string(str[i])

		if _, ok := m[char]; ok {
			m[char]++
		} else {
			m[char] = 1
		}
	}

	for i := 0; i < len(str); i++ {
		char := string(str[i])

		if m[char] == 1 {
			return i
		}
	}

	return -1
}

File: HashTable/find_first_unique_char.rs
use std::collections::HashMap;

fn main() {
  let s = "MERMAIDDERAIo";
  let index = find_first_unique_char(s.to_string());

  println!("Index of unique char is {} ", index);
}

fn find_first_unique_char(s: String) -> i32 {
  let mut hash_table = HashMap::new();
  let bytes = s.as_bytes();

  for (_, &char) in bytes.iter().enumerate() {
    let v = hash_table.entry(char).or_insert(0);
    *v += 1;
  }

  for (i, &char) in bytes.iter().enumerate() {
    let index: i32 = match hash_table.get(&char) {
      Some(v) => *v as i32,
      _ => 0,
    };

    if index == 1 {
      return i as i32;
    }
  }

  return -1;
}